### PR TITLE
[Reland] load_state_dict post hook

### PR DESCRIPTION
**Pull request by rohan-varma**
Reland of https://github.com/pytorch/pytorch/pull/76823 with fixes to call `__setstate__` for softmax/softmin/logsoftmax as per discussion with @albanD and @jbschlosser. Original description:

Implements `register_load_state_dict_post_hook` API as discussed in https://github.com/pytorch/pytorch/issues/75287.

Unittests cover:
- Ensuring hooks are called with the correct module
- Hook is called with `IncompatibleKeys` field
- If hook modifies this, load_state_dict returns the modified result

### CODE REVIEW

File: torch/nn/modules/activation.py [hunk 1]
```diff
@@ -1304,10 +1304,11 @@ def __init__(self, dim: Optional[int] = None) -> None:
         self.dim = dim
 
     def __setstate__(self, state):
-        self.__dict__.update(state)
```
Comment (rohan-varma) [inline torch/nn/modules/activation.py:1307]: @jbschlosser @albanD taking out this call since we call into super() now, and super() does `self.__dict__.update(state)` so this should be safe.

File: torch/nn/modules/activation.py [hunk 2]
```diff
@@ -1304,10 +1304,11 @@ def __init__(self, dim: Optional[int] = None) -> None:
...
         if not hasattr(self, 'dim'):
             self.dim = None
 
+        super().__setstate__(state)
```
Comment (jbschlosser) [inline torch/nn/modules/activation.py:1310]: nit: Why put this at the end? Probably doesn't matter, but I expected it at the beginning :p

File: torch/nn/modules/activation.py (hunk 1 above)
Comment (jbschlosser) [reply torch/nn/modules/activation.py:1307]: Cool, sounds good :)

File: torch/nn/modules/activation.py (hunk 2 above)
Comment (rohan-varma) [reply torch/nn/modules/activation.py:1310]: Sure, doesn't seem like it matters but moved it to the beginning

**Review by jbschlosser — APPROVED** [review]
LGTM, thanks!
